Add files via upload

juliareynolds11 opened the pull request:
Files for calculating variables over intervals

juliareynolds11 committed:
Add files via upload

rhuang10 commented on Codes/Clean Data/getAggressiveness.R:
```diff
@@ -0,0 +1,22 @@
+function(direction,
```
Interesting to know you can indeed define R function in this way. I always think the R function must be defined at this way
```R
functionName <- function(arg1, arg2) {
...
}
```
A benefit of this is that we can put multiple function definitions in one file. I wonder what is the benefit of defining function names by the file name. -- Indeed, many people do not like Matlab (and Java) which force programmer to put one function (Class/Interface in Java) in one file

juliareynolds11 replied:
As we discussed today on Skype, I keep the functions in separate files to more easily use them across multiple code. However, I can see the benefit of keeping the number of subfiles to a minimum!

rhuang10 commented on Codes/Clean Data/getAggressiveness.R:
```diff
@@ -0,0 +1,22 @@
+         ssquote,
+         indexSub,
+         string) {
+  
```
I may ask something that I fell most annoying myself. :(  The docstring for the function. At less the meaning of argument. Trust me, I hate docstring as much as other modelers. But it's essential for other people to understand what you are doing in the function. Here is the R docstring package introduction: https://cran.r-project.org/web/packages/docstring/vignettes/docstring_intro.html

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+                  include.mean = F), silent = T)
+  if (class(test) == "try-error") {
+    return(c(NA, NA, NA))
+    break
```
This is confusing. 1) Break outside of loop. 2) I don't think R will execute a statement after return in a function body.

juliareynolds11 replied:
You are correct, the "return" here is indeed redundant.

rhuang10 replied:
Are you sure return is redundant? I thought "break" is redundant. I know the "break" will break the loop of the script that is calling this function. But please be careful, because "break" without the loop inside the same function makes the function  unsafe for reusing in other script, when there is no loop the script will crash!

rhuang10 replied:
Indeed you delete break! Cool. Thank you!

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+function(indexExe, lags, price, direction, sz) {
```
The same comment as the getAggressivness.R.

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+  #"-1" denotes no constant
+  if (class(fit) == "try-error") {
+    return(c(NA, NA, NA)) #skip if model estimation is not possible
+    break
```
Again, I don't think "break" will be executed.

juliareynolds11 replied:
Agreed and removed.

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+      paste0(label, 5)
+    )
+  
+  fit <- try(lm(y ~ xmtx - 1), silent = T)
```
I understand right, $a_j^*$ and $b_j^*$ are deterministic functions of $a_j$, $b_j$, $c_j$ and $d_j$. The linear regression is a trick to compute those values. Am I right?

juliareynolds11 replied:
st (pricing errors) is a function of alpha and beta, and alpha and beta are the functions of coefficients from a VMA representation of a VAR system. The errors from the VAR are used to approximate the coefficients of the VMA. I'm not sure if this is the best way to do this, but it might be the simplest. The VAR errors are sometimes used as the starting values when estimating a VMA (see, e.g., https://faculty.chicagobooth.edu/ruey.tsay/teaching/mts/sp2011/VMA1.R or https://communities.sas.com/t5/Statistical-Procedures/R-Square-as-in-Hasbrouck-s-Review-of-Financial-Studies-1991/td-p/44948 for an example in SAS) but a few early attempts at estimating the VMA lead me to some convergence troubles. I figured that this "simpler" approach would also be faster.

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+    coefs[(dimx + 1):length(coefs)] #get rid of coefficients on lag 0
+  
+  a <- rev(coefs[seq(1, dimx * lags, dimx)]) #coefficients on v1
+  astar <- (-rev(cumsum(a)))
```
This is alpha in the equation (12) the paper, isn't it?

Also there is a typo in equation (12), beta should equal to sum of b*.

juliareynolds11 replied:
Both are good points. I have fixed the notation.

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+  a <- rev(coefs[seq(1, dimx * lags, dimx)]) #coefficients on v1
+  astar <- (-rev(cumsum(a)))
+  
+  b1 <- rev(coefs[seq(2, dimx * lags, dimx)]) #coefficients on v21
```
This is beta in the paper, isn't it?

juliareynolds11 replied:
That's right, I have fixed the notation.

rhuang10 commented on Codes/Clean Data/getHasbrouck.R:
```diff
@@ -0,0 +1,108 @@
+  st <- matrix(0, nrow(xmtx), 1)
+  
+  for (ii in 1:length(alpha)) {
+    st <- st + (alpha[ii] * xmtx[, ii])
```
Don't quite get this line. Could you explain a little bit on the dimension of alpha, xmtx and st? Thanks.

rhuang10 commented on Codes/Clean Data/getMessageType.R:
```diff
@@ -0,0 +1,62 @@
+function(condition,
```
Same comment as getAggressiveness.R

rhuang10 commented on Codes/Clean Data/getMessageType.R:
```diff
@@ -0,0 +1,62 @@
+  buySVol <- sum((condition & direction == 1) * sz)
+  sellSVol <- sum((condition & direction == -1) * sz)
+  if (buySVol + sellSVol != totalSVol) {
+    stop("Buy and Sell share volumes don't add up")
```
Indeed, you are checking whether directions are all in Set (1, -1). Am I right?

juliareynolds11 replied:
Yes, this is just a "sanity check" to make sure the buy + sell = total volume.

rhuang10 commented on Codes/Clean Data/getOrderSize.R:
```diff
@@ -0,0 +1,17 @@
+function(price, sz, indexSub, string) {
```
Same comments are in getAggressiveness.R

rhuang10 commented on Codes/Clean Data/getOrderSize.R:
```diff
@@ -0,0 +1,17 @@
+  
+  orszSVol <- sz[indexSub] #order size in terms of share volume
+  orszDVol <-
+    sz[indexSub] * price[indexSub] #order size in terms of dollar volume
```
What is the price? Is it market price or midquote or the limit price?

juliareynolds11 replied:
In this context, "price" is the price at which the order is submitted.

rhuang10 commented on Codes/Clean Data/getRealizedVolatility.R:
```diff
@@ -0,0 +1,84 @@
+function(n,
```
Same comments as in getAggressiveness.R

rhuang10 commented on Codes/Clean Data/getRealizedVolatility.R:
```diff
@@ -0,0 +1,84 @@
+      RV <- matrix(NA, 1, n)
+      
+      for (ii in 1:(length(intervaltemp) - 1)) {
+        iia <- min(which(time > intervaltemp[ii]))
```
Every time you use which, R has to scan all time data again. You want to get is the previous midquote at each interval. You can indeed create a data.table with regular time interval and do as-of join on quotes data.table.

rhuang10 commented on Codes/Clean Data/getTWSpreads.R:
```diff
@@ -0,0 +1,30 @@
+function(endSpread, spread, mid, time, intervals, q) {
```
Same comments as in getArragessiveness.R. In particular, what is q?

juliareynolds11 committed:
Update getAggressiveness.R

Added docstring description.

juliareynolds11 committed:
Update getMessageType.R

Added docstring.

juliareynolds11 committed:
Update getOrderSize.R

Added docstring

juliareynolds11 committed:
Update getRealizedVolatility.R

Added docstring. Changed calculation of subsamples to move over midquote table (much faster).

juliareynolds11 committed:
Update getTWSpreads.R

Added docstring

juliareynolds11 committed:
Update getHasbrouck.R

Added docstring. Fixed notation. Removed "breaks".

juliareynolds11 committed:
Update CalculateIntervalVariables_20200403.R

Added calculation of midquote table grid (for faster calculation of realized volatility). Added protocol to skip interval if there is an insufficient number of messages in that interval.

juliareynolds11 committed:
Update getHasbrouck.R

I realized that sometimes the estimation of the VMA will produce coefficients of NA (i.e., missing) if there are multicollinearity issues. Therefore I added a protocol to break if this occurs.
